test/test_trio_server/mock_lean.py

```diff
@@ -0,0 +1,152 @@
+    """
+    seconds: float
+
+
+@dataclass
+class LeanShouldGetRequest(LeanScriptStep):
+    """
+    Used to check that Lean gets the request it is looking for from Lean-Python interface.
+    """
+    message: Dict  # should be JSON encodable
+
+
+@dataclass
+class LeanSendsResponse(LeanScriptStep):
+    """
+    Simulate Lean's output behaviour
+    """
+    message: Dict  # should be JSON encodable
+
+
+@dataclass
+class LeanSendsBytes(LeanScriptStep):
+    """
+    Simulate Lean's output behaviour at the byte level.
+    Doesn't include new line characters unless passed in.
+    """
+    message_bytes: bytes
+
+
+@dataclass
+class LeanShouldNotGetRequest(LeanScriptStep):
+    """
+    Check that Lean has not received any requests (perhaps sent prematurely).
+    """
+    pass
+
+
+
+class MockLeanServerProcess(trio.Process):
+
+    def __init__(self, script: List[LeanScriptStep]):
+        self.stdin = trio.testing.MemorySendStream()      # a stream for mock lean to read from
+        self.stdout = trio.testing.MemoryReceiveStream()  # a stream for mock lean to write to
+        self.messages: deque[Dict] = deque()
+        self.partial_message: bytes = b""
+        self.script: List[LeanScriptStep] = script
+
+    def kill(self):
+        pass
+
+    @staticmethod
+    def parse_message(b: bytes) -> Dict:
+        return json.loads(b.decode())
+
+    @staticmethod
+    def messages_are_equal(m1: Dict, m2: Dict):
+        s1 = json.dumps(m1, sort_keys=True)
+        s2 = json.dumps(m2, sort_keys=True)
+        return s1 == s2
+
+    def collect_stdin_messages(self):
+        try:
+            data = self.stdin.get_data_nowait()
+        except trio.WouldBlock:  # no data
+            return
+
+        raw_messages = (self.partial_message + data).split(b"\n")
+        self.partial_message = raw_messages.pop()
+        self.messages.extend(self.parse_message(m) for m in raw_messages)
+
+    async def assert_message_is_received(self, message_expected):
+        await trio.sleep(0.01)
+        self.collect_stdin_messages()
+
+        assert self.messages, f"Mock Lean was expecting\n{message_expected}\nbut no messages were received."
+
+        message_received = self.messages.pop()
+        assert self.messages_are_equal(message_expected, message_received), \
+            f"Mock Lean was expecting\n{message_expected}\nbut received\n{message_received}"
+
+    async def assert_no_messages_received(self):
+        await trio.sleep(0.01)
+        self.collect_stdin_messages()
+
+        assert not self.messages, f"Mock Lean was not expecting as message, but \n{self.messages[0]}\nwas received."
+
+    def send_bytes(self, message_bytes: bytes):
+        self.stdout.put_data(message_bytes)
+
+    def send_message(self, message):
+        message_bytes = json.dumps(message).encode() + b"\n"
+        self.stdout.put_data(message_bytes)
+
+    async def follow_script(self):
+        for step in self.script:
```

Comment:
Why not defining a `run` method for each subclass of `LeanScriptStep` so that this loop body can be reduced to `step.run(self)`?

Reply:
How then would each LeanScriptStep relate to MockLeanProcess?  Would, say, LeanShouldGetRequest.run have access to the mock process?  This seems ultimately more complicated.  Here I'm thinking of LeanScriptStep as essentially an inductive data type and this is just a match statement over that data type.  (Of course lean doesn't have these concepts but that is the idea.)

Reply:
But I don't want to be too dogmatic either way, if you think that is best, we can do it.

Reply:
I indicated the relation in my message, by passing `self` to `step.run`. Now your answer exactly emphasize the issue, assuming you intended to write "python" instead of "lean" in the parenthesis. What you wrote is not python code. This is Lean (or Haskell, or Rust...) translated to python. I'll try to push a commit to this branch to show the difference.

Reply:
Yes, I meant to write Python.  Sorry, I didn't mean to imply what you said couldn't work  I agree it can.  I guess I have a resistance even in python, to classes depending on each other when it doesn't add much.  But I agree what I'm doing is already nonstandard with having a class for each step of the script.  Yes, how about you rewrite this as you think best and we can go from there.

Reply:
I tried this in 64ab19c. It does reinforce coupling between those classes, but they are very tightly related anyway. It seems to me that this solution will scale better.